### PR TITLE
Unit testing for Array and String

### PR DESCRIPTION
**Pull request by GalAster**
- **Passed:**
- [x] Array.prototype.join()
- [x] Array.prototype.concat()
- [x] String.prototype.repeat()

- **Need fix:**
- [ ] String.prototype.length
- [ ] String.prototype.concat()

### CODE REVIEW

**Comment by GalAster** [comment]
Bug: @jasonwilliams 

```ts
const s = new String('\ud834\udf06');
console.log(s.length)
```

Should be `2` but return `4`.

**Comment by GalAster** [comment]
TODO: 

- [ ] Format Array display form.

I don't know how to do this.

**Comment by GalAster** [comment]
Bug:

```ts
const hello = "Hello, ";
const world = "world! ";
hello.concat(world)
```

**Comment by GalAster** [comment]
Please review: @jasonwilliams 

1. May conflict with #69 , I use a step by step vm in unit testing.

2. CI build failed because of `The command "cargo clippy --verbose" exited with 101.`

It is caused by the code in #56, I can't modify it.

3. I tried to fix String.prototype.length, but this problem is a bit complicated.

`this_str.len()` counts UTF8 length, I changed as `this_str.chars().count()` to count unicode surrogate pair length.

But js string length is UTF16/USC2 length, still different on some special cases.

File: src/lib/js/array.rs [hunk 1]
```diff
@@ -167,7 +167,57 @@ pub fn _create(global: &Value) -> Value {
...
+
+    #[test]
+    fn concat() {
+        //TODO: array display formatter
+        let mut engine = Executor::new();
+        let init = r#"
+        let empty = [ ];
+        let one = [1];
+        "#;
+        forward(&mut engine, init);
```
Comment (jasonwilliams) [inline src/lib/js/array.rs]: Why `forward` as a name?

File: src/lib/lib.rs [hunk 2]
```diff
@@ -44,22 +45,28 @@ extern "C" {
...
 
-pub fn exec(src: &str) -> String {
+fn parser_expr(src: &str) -> Expr {
     let mut lexer = Lexer::new(src);
     lexer.lex().unwrap();
     let tokens = lexer.tokens;
+    Parser::new(tokens).parse_all().unwrap()
+}
 
+pub fn forward(engine: &mut Interpreter, src: &str) -> String {
```
Comment (jasonwilliams) [inline src/lib/lib.rs]: why `forward` as a name?

File: src/lib/js/string.rs [hunk 3]
```diff
@@ -492,10 +493,111 @@ pub fn init(global: &Value) {
...
+    fn starts_with() {
+        let mut engine = Executor::new();
+        let init = r#"
+        const empty = new String('');
+        const en = new String('english');
+        const zh = new String('中文');
+        "#;
+        forward(&mut engine, init);
+        let pass = String::from("true");
+        assert_eq!(dbg!(forward(&mut engine, "empty.startsWith('')")), pass);
```
Comment (jasonwilliams) [inline src/lib/js/string.rs]: im assuming the `dbg!()` are just for you to test?

File: src/lib/lib.rs (hunk 2 above)
Comment (jasonwilliams) [reply src/lib/lib.rs]: Would also be worth some comments above this function definition to explain what its purpose is. (compared to exec)

File: src/lib/js/string.rs [hunk 4]
```diff
@@ -492,10 +493,111 @@ pub fn init(global: &Value) {
...
+        assert_eq!(
+            dbg!(forward(&mut engine, "en.repeat(1)")),
+            String::from("english")
+        );
+        assert_eq!(
+            dbg!(forward(&mut engine, "zh.repeat(2)")),
+            String::from("中文中文")
+        );
+    }
+
```
Comment (jasonwilliams) [inline src/lib/js/string.rs]: Good tests!

File: src/lib/js/string.rs (hunk 3 above)
Comment (GalAster) [reply src/lib/js/string.rs]: Well, not necessary indeed.

File: src/lib/lib.rs (hunk 2 above)
Comment (GalAster) [reply src/lib/lib.rs]: Let me see, it was call `run_frame` in RustPython.

I'm not sure what this should be called, `forward` sounds like walk a small step, which means the step by step interpretation.

**Comment by jasonwilliams** [comment]
@GalAster you looking for me to merge this or you wanting to carry on?

**Comment by GalAster** [comment]
@jasonwilliams It's ready to merge, I'll do the rest after Array format complete.

**Pull request merged**